livre/core/pipeline/AsyncData.cpp

```
@@ -0,0 +1,154 @@
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <livre/core/pipeline/AsyncData.h>
+#include <livre/core/pipeline/PortData.h>
+#include <livre/core/pipeline/Future.h>
+
+#include <lunchbox/log.h>
+
+#include <boost/thread/future.hpp>
+
+namespace livre
+{
+
+typedef boost::shared_future< PortDataPtr > ConstPortDataFuture;
+typedef boost::promise< PortDataPtr > ConstPortDataPromise;
+typedef std::vector< ConstPortDataFuture > ConstPortDataFutures;
+
+struct AsyncData::Impl
+{
+    Impl( const DataInfo& dataInfo )
+        : _future( _promise.get_future( ))
+        , _dataInfo( dataInfo )
+    {}
+
+    ~Impl()
+    {
+        set( PortDataPtr( ));
+    }
+
+    PortDataPtr get( const std::type_index& dataType ) const
+    {
+        const PortDataPtr& data = _future.get();
+
+        if( !data )
+            LBTHROW( std::runtime_error( "Returns empty data" ));
+
+        if( data->dataType != dataType )
+            LBTHROW( std::runtime_error( "Types does not match on get value"));
+
+        return data;
+    }
+
+    void set( const PortDataPtr& data )
+    {
+        if( data )
+        {
+            if( _dataInfo.second != data->dataType )
+                LBTHROW( std::runtime_error( "Types does not match on set value"));
+        }
+
+        try
+        {
+            _promise.set_value( data );
+        }
+        catch( const boost::promise_already_satisfied& )
```

Why do you fail silently?